CI failure looks related;

```
=== FAIL: arm64.integration-cli TestDockerDaemonSuite/TestDaemonNoSpaceLeftOnDeviceError (0.21s)
    docker_cli_daemon_test.go:1118: assertion failed: error is not nil: error from registry: unknown error - unrecognized manifest schema version 0
    --- FAIL: TestDockerDaemonSuite/TestDaemonNoSpaceLeftOnDeviceError (0.21s)
```

--- a/integration-cli/docker_cli_daemon_test.go
+++ b/integration-cli/docker_cli_daemon_test.go
@@ -26,6 +26,7 @@ import (
 	"github.com/cloudflare/cfssl/helpers"
 	cerrdefs "github.com/containerd/errdefs"
 	"github.com/creack/pty"
+	archive "github.com/moby/go-archive"
 	"github.com/moby/moby/client"
 	"github.com/moby/moby/v2/daemon/pkg/opts"
 	"github.com/moby/moby/v2/integration-cli/checker"
@@ -34,6 +35,8 @@ import (
 	"github.com/moby/moby/v2/integration-cli/daemon"
 	"github.com/moby/moby/v2/internal/testutil"
 	testdaemon "github.com/moby/moby/v2/internal/testutil/daemon"
+	"github.com/moby/moby/v2/internal/testutil/registry"
+	"github.com/moby/moby/v2/internal/testutil/specialimage"
 	"github.com/moby/sys/mount"
 	"golang.org/x/sys/unix"
 	"gotest.tools/v3/assert"
@@ -1077,13 +1080,46 @@ func (s *DockerDaemonSuite) TestBridgeIPIsExcludedFromAllocatorPool(c *testing.T
 
 // Test daemon for no space left on device error
 func (s *DockerDaemonSuite) TestDaemonNoSpaceLeftOnDeviceError(c *testing.T) {
-	testRequires(c, testEnv.IsLocalDaemon, DaemonIsLinux, Network)
+	testRequires(c, testEnv.IsLocalDaemon, DaemonIsLinux, RegistryHosting)
 
 	testDir, err := os.MkdirTemp("", "no-space-left-on-device-test")
 	assert.NilError(c, err)
 	defer os.RemoveAll(testDir)
 	assert.NilError(c, mount.MakeRShared(testDir))
 	defer mount.Unmount(testDir)
+
+	// Start a local registry to avoid pulling from Docker Hub (rate limits,
+	// transient network errors). 127.0.0.0/8 is auto-insecure for both the
+	// host and test daemons, so no extra flags are needed.
+	reg := registry.NewV2(c)
+	reg.WaitReady(c)
+	defer reg.Close()
+
+	// Create a synthetic 4 MiB image using the host daemon and push it to
+	// the local registry. The test daemon's data-root is only 3 MiB, so
+	// we use the host daemon for image setup. The image is removed from the
+	// host immediately after pushing — it is not needed locally after that.
+	ctx := testutil.GetContext(c)
+	apiClient := testEnv.APIClient()
+
+	const bigImageRef = registry.DefaultURL + "/testbig:latest"
+	imageTempDir := c.TempDir()
+	_, err = specialimage.MultiLayerCustom(imageTempDir, bigImageRef, []specialimage.SingleFileLayer{
+		{Name: "bigfile", Content: make([]byte, 4<<20)},
+	})
+	assert.NilError(c, err)
+	tar, err := archive.TarWithOptions(imageTempDir, &archive.TarOptions{})
+	assert.NilError(c, err)
+	loadResp, err := apiClient.ImageLoad(ctx, tar, client.ImageLoadWithQuiet(true))
+	assert.NilError(c, err)
+	_, err = io.Copy(io.Discard, loadResp)
+	assert.NilError(c, err)
+	_ = loadResp.Close()
+	pushResp, err := apiClient.ImagePush(ctx, bigImageRef, client.ImagePushOptions{RegistryAuth: "{}"})
+	assert.NilError(c, err)
+	assert.NilError(c, pushResp.Wait(ctx))
+	_, err = apiClient.ImageRemove(ctx, bigImageRef, client.ImageRemoveOptions{Force: true})
+	assert.NilError(c, err)
 
 	// create a 3MiB image (with a 2MiB ext4 fs) and mount it as storage root
 	storageFS := filepath.Join(testDir, "testfs.img")
@@ -1111,8 +1147,8 @@ func (s *DockerDaemonSuite) TestDaemonNoSpaceLeftOnDeviceError(c *testing.T) {
 	)
 	defer s.d.Stop(c)
 
-	// pull a repository large enough to overfill the mounted filesystem
-	pullOut, err := s.d.Cmd("pull", "debian:trixie-slim")
+	// Pull the large image into the tiny data-root to trigger ENOSPC.
+	pullOut, err := s.d.Cmd("pull", bigImageRef)
 	assert.Check(c, err != nil)
 	assert.Check(c, is.Contains(pullOut, "no space left on device"))
 }

--- a/internal/testutil/specialimage/labeled.go
+++ b/internal/testutil/specialimage/labeled.go
@@ -3,6 +3,7 @@ package specialimage
 import (
 	"github.com/containerd/platforms"
 	"github.com/distribution/reference"
+	"github.com/opencontainers/image-spec/specs-go"
 	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
 )
 
@@ -28,6 +29,7 @@ func Labeled(dir string, imageRef string, labels map[string]string) (*ocispec.In
 	}
 
 	manifest := ocispec.Manifest{
+		Versioned: specs.Versioned{SchemaVersion: 2},
 		MediaType: ocispec.MediaTypeImageManifest,
 		Config:    configDesc,
 		Layers:    []ocispec.Descriptor{},

--- a/internal/testutil/specialimage/multilayer.go
+++ b/internal/testutil/specialimage/multilayer.go
@@ -60,6 +60,7 @@ func MultiLayerCustom(dir string, imageRef string, layers []SingleFileLayer) (*o
 	}
 
 	manifest := ocispec.Manifest{
+		Versioned: specs.Versioned{SchemaVersion: 2},
 		MediaType: ocispec.MediaTypeImageManifest,
 		Config:    configDesc,
 		Layers:    layerDescs,

--- a/internal/testutil/specialimage/random.go
+++ b/internal/testutil/specialimage/random.go
@@ -9,6 +9,7 @@ import (
 
 	"github.com/distribution/reference"
 	"github.com/opencontainers/go-digest"
+	"github.com/opencontainers/image-spec/specs-go"
 	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
 )
 
@@ -43,6 +44,7 @@ func RandomSinglePlatform(dir string, platform ocispec.Platform, source rand.Sou
 	}
 
 	manifest := ocispec.Manifest{
+		Versioned: specs.Versioned{SchemaVersion: 2},
 		MediaType: ocispec.MediaTypeImageManifest,
 		Config:    configDesc,
 		Layers:    layers,

--- a/internal/testutil/specialimage/textplain.go
+++ b/internal/testutil/specialimage/textplain.go
@@ -4,6 +4,7 @@ import (
 	"strings"
 
 	"github.com/distribution/reference"
+	"github.com/opencontainers/image-spec/specs-go"
 	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
 )
 
@@ -23,7 +24,9 @@ func TextPlain(dir string) (*ocispec.Index, error) {
 	configDesc.MediaType = "application/vnd.oci.empty.v1+json"
 
 	desc, err := writeJsonBlob(dir, ocispec.MediaTypeImageManifest, ocispec.Manifest{
-		Config: configDesc,
+		Versioned: specs.Versioned{SchemaVersion: 2},
+		MediaType: ocispec.MediaTypeImageManifest,
+		Config:    configDesc,
 		Layers: []ocispec.Descriptor{
 			emptyJsonDesc,
 		},

--- a/internal/testutil/specialimage/twoplatform.go
+++ b/internal/testutil/specialimage/twoplatform.go
@@ -63,6 +63,7 @@ func oneLayerPlatformManifest(dir string, platform ocispec.Platform, f FileInLay
 	}
 
 	manifestDesc, err := writeJsonBlob(dir, ocispec.MediaTypeImageManifest, ocispec.Manifest{
+		Versioned: specs.Versioned{SchemaVersion: 2},
 		MediaType: ocispec.MediaTypeImageManifest,
 		Config:    configDesc,
 		Layers:    []ocispec.Descriptor{layerDesc},

